In [13]:
import pandas as pd

In [14]:
df = pd.read_csv('/content/dataset_med.csv')

## Data Overview and Cleaning



In [15]:
print("First 5 rows of the DataFrame:")
print(df.head())

print("\nConcise summary of the DataFrame:")
df.info()

print("\nDescriptive statistics for numerical columns:")
print(df.describe())

print("\nNumber of missing values per column:")
print(df.isnull().sum())

First 5 rows of the DataFrame:
   id   age  gender      country diagnosis_date cancer_stage family_history  \
0   1  64.0    Male       Sweden     2016-04-05      Stage I            Yes   
1   2  50.0  Female  Netherlands     2023-04-20    Stage III            Yes   
2   3  65.0  Female      Hungary     2023-04-05    Stage III            Yes   
3   4  51.0  Female      Belgium     2016-02-05      Stage I             No   
4   5  37.0    Male   Luxembourg     2023-11-29      Stage I             No   

   smoking_status   bmi  cholesterol_level  hypertension  asthma  cirrhosis  \
0  Passive Smoker  29.4                199             0       0          1   
1  Passive Smoker  41.2                280             1       1          0   
2   Former Smoker  44.0                268             1       1          0   
3  Passive Smoker  43.0                241             1       1          0   
4  Passive Smoker  19.7                178             0       0          0   

   other_cancer tre

## Feature Engineering and Target Preparation

Prepare the target variable 'survived' and engineer new features from existing columns. This includes handling missing values in the target variable, converting date columns into a numerical duration, and transforming categorical features into numerical ones where appropriate.


**Reasoning**:
The first step is to handle missing values in the target variable 'survived' by removing rows where it is null, as specified in the instructions. This ensures the target variable is complete for subsequent steps.



In [16]:
initial_rows = df.shape[0]
df.dropna(subset=['survived'], inplace=True)
rows_after_drop = df.shape[0]
print(f"Removed {initial_rows - rows_after_drop} rows with missing 'survived' values.")
print(f"DataFrame now has {rows_after_drop} rows.")

Removed 0 rows with missing 'survived' values.
DataFrame now has 890000 rows.


**Reasoning**:
Now that missing values in the target variable are handled, I will convert the date columns to datetime objects, calculate the treatment duration, and display the first few rows with the new feature.



In [17]:
df['diagnosis_date'] = pd.to_datetime(df['diagnosis_date'])
df['end_treatment_date'] = pd.to_datetime(df['end_treatment_date'])
df['treatment_duration'] = (df['end_treatment_date'] - df['diagnosis_date']).dt.days

print("DataFrame head with new 'treatment_duration' feature:")
print(df.head())

DataFrame head with new 'treatment_duration' feature:
   id   age  gender      country diagnosis_date cancer_stage family_history  \
0   1  64.0    Male       Sweden     2016-04-05      Stage I            Yes   
1   2  50.0  Female  Netherlands     2023-04-20    Stage III            Yes   
2   3  65.0  Female      Hungary     2023-04-05    Stage III            Yes   
3   4  51.0  Female      Belgium     2016-02-05      Stage I             No   
4   5  37.0    Male   Luxembourg     2023-11-29      Stage I             No   

   smoking_status   bmi  cholesterol_level  hypertension  asthma  cirrhosis  \
0  Passive Smoker  29.4                199             0       0          1   
1  Passive Smoker  41.2                280             1       1          0   
2   Former Smoker  44.0                268             1       1          0   
3  Passive Smoker  43.0                241             1       1          0   
4  Passive Smoker  19.7                178             0       0          0 

**Reasoning**:
The next step is to convert the 'family_history' column from categorical ('Yes', 'No') to numerical (1, 0) as specified in the instructions. This will prepare it for machine learning models.



In [18]:
df['family_history'] = df['family_history'].map({'Yes': 1, 'No': 0})
print("DataFrame head with numerical 'family_history' feature:")
print(df.head())

DataFrame head with numerical 'family_history' feature:
   id   age  gender      country diagnosis_date cancer_stage  family_history  \
0   1  64.0    Male       Sweden     2016-04-05      Stage I               1   
1   2  50.0  Female  Netherlands     2023-04-20    Stage III               1   
2   3  65.0  Female      Hungary     2023-04-05    Stage III               1   
3   4  51.0  Female      Belgium     2016-02-05      Stage I               0   
4   5  37.0    Male   Luxembourg     2023-11-29      Stage I               0   

   smoking_status   bmi  cholesterol_level  hypertension  asthma  cirrhosis  \
0  Passive Smoker  29.4                199             0       0          1   
1  Passive Smoker  41.2                280             1       1          0   
2   Former Smoker  44.0                268             1       1          0   
3  Passive Smoker  43.0                241             1       1          0   
4  Passive Smoker  19.7                178             0       0    

**Reasoning**:
As specified in the instructions, the 'id', 'diagnosis_date', and 'end_treatment_date' columns should be dropped. The 'id' column is not useful for prediction, and the date columns have been used to create the 'treatment_duration' feature.



In [19]:
df.drop(columns=['id', 'diagnosis_date', 'end_treatment_date'], inplace=True)
print("DataFrame head after dropping unnecessary columns:")
print(df.head())

print("\nDataFrame info after dropping columns:")
df.info()

DataFrame head after dropping unnecessary columns:
    age  gender      country cancer_stage  family_history  smoking_status  \
0  64.0    Male       Sweden      Stage I               1  Passive Smoker   
1  50.0  Female  Netherlands    Stage III               1  Passive Smoker   
2  65.0  Female      Hungary    Stage III               1   Former Smoker   
3  51.0  Female      Belgium      Stage I               0  Passive Smoker   
4  37.0    Male   Luxembourg      Stage I               0  Passive Smoker   

    bmi  cholesterol_level  hypertension  asthma  cirrhosis  other_cancer  \
0  29.4                199             0       0          1             0   
1  41.2                280             1       1          0             0   
2  44.0                268             1       1          0             0   
3  43.0                241             1       1          0             0   
4  19.7                178             0       0          0             0   

  treatment_type  survi

**Reasoning**:
The subtask requires transforming categorical features into numerical ones. I will identify the remaining object-type columns that represent categorical variables, which are 'gender', 'country', 'cancer_stage', 'smoking_status', and 'treatment_type'. These columns will be one-hot encoded to convert them into a numerical format suitable for machine learning models.



In [20]:
categorical_cols = ['gender', 'country', 'cancer_stage', 'smoking_status', 'treatment_type']
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

print("DataFrame head after one-hot encoding categorical features:")
print(df.head())

print("\nDataFrame info after one-hot encoding:")
df.info()

DataFrame head after one-hot encoding categorical features:
    age  family_history   bmi  cholesterol_level  hypertension  asthma  \
0  64.0               1  29.4                199             0       0   
1  50.0               1  41.2                280             1       1   
2  65.0               1  44.0                268             1       1   
3  51.0               0  43.0                241             1       1   
4  37.0               0  19.7                178             0       0   

   cirrhosis  other_cancer  survived  treatment_duration  ...  country_Sweden  \
0          1             0         0                 523  ...            True   
1          0             0         1                 424  ...           False   
2          0             0         0                 370  ...           False   
3          0             0         0                 443  ...           False   
4          0             0         0                 406  ...           False   

   cance

## Data Preprocessing for Machine Learning

Scale numerical features to prepare the data for model training.


The next step is to scale the specified numerical features ('age', 'bmi', 'cholesterol_level', and 'treatment_duration') using `StandardScaler` to prepare them for model training, as instructed.



In [21]:
from sklearn.preprocessing import StandardScaler

numerical_cols = ['age', 'bmi', 'cholesterol_level', 'treatment_duration']

scaler = StandardScaler()
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

print("DataFrame head with scaled numerical features:")
print(df.head())

print("\nDescriptive statistics for scaled numerical columns:")
print(df[numerical_cols].describe())

DataFrame head with scaled numerical features:
        age  family_history       bmi  cholesterol_level  hypertension  \
0  0.899796               1 -0.130748          -0.797424             0   
1 -0.500977               1  1.279296           1.067550             1   
2  0.999851               1  1.613882           0.791257             1   
3 -0.400922               0  1.494387           0.169599             1   
4 -1.801696               0 -1.289852          -1.280936             0   

   asthma  cirrhosis  other_cancer  survived  treatment_duration  ...  \
0       0          1             0         0            0.465906  ...   
1       1          0             0         1           -0.244658  ...   
2       1          0             0         0           -0.632238  ...   
3       1          0             0         0           -0.108287  ...   
4       0          0             0         0           -0.373851  ...   

   country_Sweden  cancer_stage_Stage II  cancer_stage_Stage III  \
0

## Model Training and Evaluation

Split the preprocessed dataset into training and testing sets. Then, train a suitable machine learning classification model to predict the 'survived' status of patients.

Evaluate the performance of the trained model using appropriate classification metrics such as accuracy, precision, recall, and F1-score to assess its predictive capability.



In [40]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Separate features (X) and target (y)
X = df.drop('survived', axis=1)
y = df['survived']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# --- Train and Evaluate Logistic Regression ---
model = LogisticRegression(max_iter=1000, random_state=42, class_weight='balanced')
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# --- Train and Evaluate RandomForestClassifier ---
rf_model = RandomForestClassifier(class_weight='balanced', random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
precision_rf = precision_score(y_test, y_pred_rf)
recall_rf = recall_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf)

# --- Train and Evaluate Decision Tree Classifier ---
dt_model = DecisionTreeClassifier(class_weight='balanced', random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
precision_dt = precision_score(y_test, y_pred_dt)
recall_dt = recall_score(y_test, y_pred_dt)
f1_dt = f1_score(y_test, y_pred_dt)

# --- Train and Evaluate XGBoost Classifier ---
neg_count = y_train.value_counts()[0]
pos_count = y_train.value_counts()[1]
scale_pos_weight_value = neg_count / pos_count
xgb_model = XGBClassifier(
    objective='binary:logistic',
    eval_metric='logloss',
    scale_pos_weight=scale_pos_weight_value,
    random_state=42
)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
precision_xgb = precision_score(y_test, y_pred_xgb)
recall_xgb = recall_score(y_test, y_pred_xgb)
f1_xgb = f1_score(y_test, y_pred_xgb)

# --- Train and Evaluate Linear SVM Classifier ---
svc_model = LinearSVC(class_weight='balanced', random_state=42, dual=False, max_iter=2000)
svc_model.fit(X_train, y_train)
y_pred_svc = svc_model.predict(X_test)
accuracy_svc = accuracy_score(y_test, y_pred_svc)
precision_svc = precision_score(y_test, y_pred_svc)
recall_svc = recall_score(y_test, y_pred_svc)
f1_svc = f1_score(y_test, y_pred_svc)

# --- Print Comparison Table ---
print("\n--- All Model Performance Comparison ---")
print(f"Metric         | Logistic Reg. | RandomForest | Decision Tree | XGBoost       | Linear SVM")
print(f"---------------|---------------|--------------|---------------|---------------|------------")
print(f"Accuracy       | {accuracy:<13.4f} | {accuracy_rf:<12.4f} | {accuracy_dt:<13.4f} | {accuracy_xgb:<13.4f} | {accuracy_svc:<10.4f}")
print(f"Precision      | {precision:<13.4f} | {precision_rf:<12.4f} | {precision_dt:<13.4f} | {precision_xgb:<13.4f} | {precision_svc:<10.4f}")
print(f"Recall         | {recall:<13.4f} | {recall_rf:<12.4f} | {recall_dt:<13.4f} | {recall_xgb:<13.4f} | {recall_svc:<10.4f}")
print(f"F1-Score       | {f1:<13.4f} | {f1_rf:<12.4f} | {f1_dt:<13.4f} | {f1_xgb:<13.4f} | {f1_svc:<10.4f}")


--- All Model Performance Comparison ---
Metric         | Logistic Reg. | RandomForest | Decision Tree | XGBoost       | Linear SVM
---------------|---------------|--------------|---------------|---------------|------------
Accuracy       | 0.4860        | 0.7794       | 0.6477        | 0.5134        | 0.4905    
Precision      | 0.2201        | 0.1776       | 0.2181        | 0.2221        | 0.2200    
Recall         | 0.5245        | 0.0005       | 0.2319        | 0.4832        | 0.5159    
F1-Score       | 0.3101        | 0.0010       | 0.2248        | 0.3043        | 0.3084    


## Final Task

Compare the performance of all trained models (Logistic Regression, RandomForestClassifier, Decision Tree, XGBoost, and SVM) across accuracy, precision, recall, and F1-score. Determine which model is most suitable for predicting patient survival, considering the class imbalance, and summarize the key insights from the overall analysis.


## Summary:

The most suitable model for predicting patient survival, considering the class imbalance, is the **Logistic Regression** model. It achieved the highest F1-score (0.3101) and recall (0.5245) among all tested models, indicating a better balance between identifying positive cases and minimizing false negatives for the minority class (survival). While other models like RandomForest achieved higher overall accuracy, their extremely low recall for the positive class makes them less suitable for this specific task where correctly identifying survival is crucial.

### Data Analysis Key Findings
*   **Model Performance Overview**: Five different classification models were trained and evaluated on the preprocessed data, all incorporating strategies to handle class imbalance.
*   **Logistic Regression**: Achieved an accuracy of 0.4860, precision of 0.2201, recall of 0.5245, and an F1-score of 0.3101. It demonstrated the highest recall and F1-score for the minority class among all models.
*   **RandomForestClassifier**: Showed the highest accuracy at 0.7794 but performed very poorly on the minority class with an extremely low recall of 0.0005 and an F1-score of 0.0010. This indicates it largely failed to predict positive survival cases.
*   **Decision Tree Classifier**: Exhibited an accuracy of 0.6477, precision of 0.2181, recall of 0.2319, and an F1-score of 0.2248. Its performance on the minority class was moderate but lower than Logistic Regression, XGBoost, and Linear SVM.
*   **XGBoost Classifier**: Achieved an accuracy of 0.5134, precision of 0.2221, recall of 0.4832, and an F1-score of 0.3043. It provided a competitive balance of precision and recall for the minority class.
*   **Linear SVM**: Recorded an accuracy of 0.4905, precision of 0.2200, recall of 0.5159, and an F1-score of 0.3084. Its performance was comparable to Logistic Regression and XGBoost in terms of minority class prediction.
*   **Class Imbalance Handling**: The use of `class_weight='balanced'` for Logistic Regression, Decision Tree, RandomForest, and Linear SVM, and `scale_pos_weight` for XGBoost, was crucial to encourage models to pay more attention to the minority class.


